README.md
![Build Status](https://travis-ci.com/alsiola/pm2-explorer.svg?branch=master)

# PM2 Explorer README

View and interact with currently running PM2 processes within VSCode.  This extension can be installed from within VSCode by searching for `pm2` in the extensions view, or via the [Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=alex-young.pm2-explorer)

## Features

Adds a PM2 Explorer view that lists current PM2 processes and important details about them. The list is updated at a configurable interval (default 1000ms).

Allows interaction with individual running processes: 

 * Start
 * Stop
 * Restart
 * View Logs (in integrated terminal)

And with *all* running processes:

 * Reload all
 * View all logs
 * Start all processes
 * Stop all processes

## Requirements

PM2 should be installed globally on your machine.

## Extension Settings

 The following configuration settings are provided by PM2 explorer:

```
// How frequently (in milliseconds) should the process list update
"pm2Explorer.refreshIntervalMs": 1000
```

## Known Issues

No known issues

## Release Notes

### 0.0.7

* Fix commands when tasks have whitespace
* Dependency vulnerability updates

### 0.0.6

* Restore accidentally removed container icon

### 0.0.5

* Re-publish broken CI deployment

### 0.0.4
* Improved process icons
* Configurable refresh interval
* Added start all and stop all commands
* Refactor and re-organisation

### 0.0.3

* Visual representation of process state

### 0.0.2

* Fix errors in readme
* Add logo

### 0.0.1

* Initial release of PM2 Explorer


CHANGELOG.md
# Change Log

### 0.0.7

    - Fix commands when tasks have whitespace
    - Dependency vulnerability updates

### 0.0.6

    - Restore accidentally removed container icon

### 0.0.5

    - Re-publish broken CI deployment

### 0.0.4

    - Improved process icons
    - Configurable refresh interval
    - Added start all and stop all commands
    - Refactor and re-organisation

### 0.0.3

    - Visual representation of process state

### 0.0.2

    - Fix errors in readme
    - Add logo

## 0.0.1

    - Initial release


src/test/util.test.ts
import * as assert from "assert";
import { millisecondsToReadable } from "../util";

const SECONDS = 1000;
const MINUTES = 60 * SECONDS;
const HOURS = 60 * MINUTES;
const DAYS = 24 * HOURS;
const WEEKS = 7 * DAYS;

suite("Util", () => {
    suite("millisecondsToReadable", () => {
        test("converts to milliseconds when less than one second", () => {
            const actual = millisecondsToReadable(5);

            const expected = "5 ms";

            assert.equal(actual, expected);
        });

        test("converts to seconds when one second to one minute", () => {
            const actual = millisecondsToReadable(5 * SECONDS);

            const expected = "5 seconds";

            assert.equal(actual, expected);
        });

        test("converts to minutes when one minute to one hour", () => {
            const actual = millisecondsToReadable(5 * MINUTES);

            const expected = "5 minutes";

            assert.equal(actual, expected);
        });

        test("converts to hours when one hour to one day", () => {
            const actual = millisecondsToReadable(5 * HOURS);

            const expected = "5 hours";

            assert.equal(actual, expected);
        });

        test("converts to days when one day to one week", () => {
            const actual = millisecondsToReadable(5 * DAYS);

            const expected = "5 days";

            assert.equal(actual, expected);
        });

        test("converts to weeks when over one week", () => {
            const actual = millisecondsToReadable(5 * WEEKS);

            const expected = "5 weeks";

            assert.equal(actual, expected);
        });
    });
});


src/model/pm2.ts
import * as nodePm2 from "pm2";
import * as vscode from "vscode";
import * as util from "../util";
import { Process } from "./process";

export class PM2
    implements vscode.TreeDataProvider<vscode.TreeItem>, vscode.Disposable {
    private _onDidChangeTreeData: vscode.EventEmitter<
        Process | undefined
    > = new vscode.EventEmitter<Process | undefined>();

    readonly onDidChangeTreeData: vscode.Event<Process | undefined> = this
        ._onDidChangeTreeData.event;

    private _pm2!: Promise<typeof nodePm2>;
    private _processes: nodePm2.ProcessDescription[] = [];
    private _isRefreshing = false;
    private _isDisposed = false;

    constructor(
        private _context: vscode.ExtensionContext,
        private readonly _config: vscode.WorkspaceConfiguration
    ) {
        console.log("Initialising PM2");
        this._pm2 = new Promise(resolve => {
            nodePm2.connect(util.errCallback(
                () => {
                    this.setRefreshInterval();
                    resolve(nodePm2);
                },
                err => util.showErr("Could not connect to PM2")
            ) as any);
        });
    }

    private setRefreshInterval() {
        util.refresher({
            fn: () => {
                this._isRefreshing = true;
                this.listProcesses().then(() => {
                    this._isRefreshing = false;
                });
            },
            interval: () =>
                this._config.get<number>("refreshIntervalMs") || 1000,
            skipIf: () => this._isRefreshing || this._isDisposed
        })();
    }

    dispose() {
        this._isDisposed = true;
    }

    logs(process?: nodePm2.ProcessDescription) {
        const terminal = vscode.window.createTerminal("pm2");
        terminal.sendText(
            `pm2 logs '${process && process.name ? process.name : ""}'`
        );
        terminal.show();
    }

    reloadAll() {
        this._pm2.then(pm2 => {
            pm2.reload("all", util.errCallback());
        });
    }

    startAll() {
        this._pm2.then(pm2 => {
            pm2.start({}, util.errCallback());
        });
    }

    stopAll() {
        this._pm2.then(pm2 => {
            pm2.stop("all", util.errCallback());
        });
    }

    reload(process: nodePm2.ProcessDescription) {
        this._pm2.then(pm2 => {
            pm2.reload(
                process.name!,
                util.errCallback(() => {
                    util.showMsg(`Reloaded process ${process.name}`);
                })
            );
        });
    }

    listProcesses() {
        console.log("Retrieving PM2 processes");
        return this._pm2.then(pm2 => {
            return new Promise(resolve =>
                pm2.list(
                    util.errCallback(processes => {
                        resolve();
                        console.log(
                            `Retrieved ${processes.length} PM2 processes`
                        );
                        this._processes = processes;
                        this._onDidChangeTreeData.fire();
                    }, resolve)
                )
            );
        });
    }

    refresh() {
        this._processes = [];
        this._onDidChangeTreeData.fire();
        this.listProcesses();
    }

    getTreeItem(element: Process): vscode.TreeItem | Thenable<vscode.TreeItem> {
        return element;
    }

    getChildren(
        element?: Process | undefined
    ): vscode.ProviderResult<vscode.TreeItem[]> {
        if (!element) {
            return this._processes.map(
                process => new Process(process, this._pm2, this._context)
            );
        }
        return element.children;
    }
}


src/model/process.ts
import * as path from "path";
import * as nodePm2 from "pm2";
import * as vscode from "vscode";
import { errCallback, showMsg, millisecondsToReadable } from "../util";

const getProcessLabel = (process: nodePm2.ProcessDescription): string => {
    return process.name!;
};

export class Process extends vscode.TreeItem {
    public readonly label?: string;
    public readonly contextValue = "pm2-process";

    constructor(
        public readonly process: nodePm2.ProcessDescription,
        private readonly _pm2: Promise<typeof nodePm2>,
        private _context: vscode.ExtensionContext
    ) {
        super(
            getProcessLabel(process),
            vscode.TreeItemCollapsibleState.Collapsed
        );
        this.label = getProcessLabel(process);
    }

    start() {
        this._pm2.then(pm2 =>
            pm2.start(
                {
                    name: this.process.name!
                },
                errCallback(() =>
                    showMsg(`Started process ${this.process.name}`)
                )
            )
        );
    }

    stop() {
        this._pm2.then(pm2 =>
            pm2.stop(
                this.process.name!,
                errCallback(() =>
                    showMsg(`Stopped process ${this.process.name}`)
                )
            )
        );
    }

    get iconPath(): string | undefined {
        const { status } = this.process.pm2_env!;

        switch (status) {
            case "errored":
            case "online":
            case "stopped":
                return this._context.asAbsolutePath(
                    path.join("resources", `${status}.png`)
                );
            case "launching":
                return this._context.asAbsolutePath(
                    path.join("resources", `${status}.gif`)
                );
        }
    }

    get tooltip(): string {
        return `Name: ${this.label}
PID: ${this.process.pid}`;
    }

    get children(): vscode.TreeItem[] {
        return [
            new vscode.TreeItem(`Status: ${this.process.pm2_env!.status}`),
            new vscode.TreeItem(
                `Instances: ${this.process.pm2_env!.instances}`
            ),
            new vscode.TreeItem(
                `Uptime: ${millisecondsToReadable(
                    Date.now() - (this.process.pm2_env!.pm_uptime || 0)
                )}`
            ),
            new vscode.TreeItem(
                `Unstable restarts: ${this.process.pm2_env!.unstable_restarts}`
            )
        ];
    }
}


src/contributions/commands.ts
import * as vscode from "vscode";
import { PM2, Process } from "../model";

export const registerCommands = (pm2: PM2) => {
    vscode.commands.registerCommand("pm2.reload", (item: Process) => {
        pm2.reload(item.process);
    });

    vscode.commands.registerCommand("pm2.logs", (item: Process) => {
        pm2.logs(item.process);
    });

    vscode.commands.registerCommand("pm2.start", (item: Process) => {
        item.start();
    });

    vscode.commands.registerCommand("pm2.stop", (item: Process) => {
        item.stop();
    });

    vscode.commands.registerCommand("pm2.monit", () => {
        pm2.monit();
    });

    vscode.commands.registerCommand("pm2.flushLogs", () => {
        pm2.flushLogs();
    });

    vscode.commands.registerCommand("pm2.logsAll", () => {
        pm2.logs();
    });

    vscode.commands.registerCommand("pm2.startAll", () => {
        pm2.startAll();
    });

    vscode.commands.registerCommand("pm2.stopAll", () => {
        pm2.stopAll();
    });

    vscode.commands.registerCommand("pm2.reloadAll", () => {
        pm2.reloadAll();
    });

    vscode.commands.registerCommand("pm2.refresh", () => {
        pm2.refresh();
    });
};


src/util.ts
import * as vscode from "vscode";

export const errCallback = <T>(
    cb?: (a: T) => void,
    reject?: ((err: any) => void)
) => (err: Error, res: T) => {
    if (err) {
        console.error(err);
        showErr(`PM2 Error:
err.message`);
        return reject && reject(err);
    }
    return cb && cb(res);
};

export const showMsg = (msg: string) => {
    vscode.window.showInformationMessage(`PM2 Explorer: ${msg}`);
};

export const showErr = (msg: string) =>
    vscode.window.showErrorMessage(`PM2 Explorer: ${msg}`);

export const refresher = ({
    fn,
    interval,
    skipIf
}: {
    fn: () => void;
    interval: () => number;
    skipIf: () => boolean;
}) => () => {
    !skipIf() && fn();
    return setTimeout(refresher({ fn, interval, skipIf }), interval());
};

export const sendTerminalCommand = (command: string, show: boolean = true) => {
    const terminal = vscode.window.createTerminal("pm2");
    terminal.sendText(command);
    if (show) terminal.show();
};

const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;
const ONE_HOUR = 60 * ONE_MINUTE;
const ONE_DAY = 24 * ONE_HOUR;
const ONE_WEEK = 7 * ONE_DAY;

const convertMilliseconds = (
    milliseconds: number,
    interval: number,
    suffix: string
) => {
    const quantity = Math.floor(milliseconds / interval);

    return `${quantity} ${suffix}`;
};

export const millisecondsToReadable = (milliseconds: number) => {
    if (milliseconds > ONE_WEEK) {
        return convertMilliseconds(milliseconds, ONE_WEEK, "weeks");
    }

    if (milliseconds > ONE_DAY) {
        return convertMilliseconds(milliseconds, ONE_DAY, "days");
    }

    if (milliseconds > ONE_HOUR) {
        return convertMilliseconds(milliseconds, ONE_HOUR, "hours");
    }

    if (milliseconds > ONE_MINUTE) {
        return convertMilliseconds(milliseconds, ONE_MINUTE, "minutes");
    }

    if (milliseconds > ONE_SECOND) {
        return convertMilliseconds(milliseconds, ONE_SECOND, "seconds");
    }

    return `${milliseconds} ms`;
};
